File: packages/web/src/algorithms/parseSequences.ts
export function sanitizeSequence(seq: string) {
  // Trim contiguous Ns in the beginning and end
  return seq // .replace(/N+$/g, '').replace(/^N+/g, '')
}

export function addSequence(
  currentSeq: string,
  currentSeqName: string,
  seqs: Record<string, string>,
  allNames: string[],
) {
  if (currentSeqName === '') {
    // eslint-disable-next-line no-param-reassign
    currentSeqName = 'Untitled'
  }

  let nameCount = 0
  for (let i = 0; i < allNames.length; ++i) {
    if (allNames[i] === currentSeqName) {
      ++nameCount
    }
  }

  let suffix = ''
  if (nameCount) {
    suffix = ` (${nameCount})`
  }

  allNames.push(currentSeqName)
  seqs[currentSeqName + suffix] = sanitizeSequence(currentSeq)
}

export function parseSequences(input: string) {
  // NOTE: This should protect from parsing failures when file is using DOS- and old Mac-style newlines
  const content = input.replace(/\r\n/g, '\n').replace(/\r/g, '\n')
  const lines = content.split('\n')
  let currentSeqName = ''
  let currentSeq = ''
  const seqs: Record<string, string> = {}
  const seqNames: string[] = []

  for (let i = 0; i < lines.length; ++i) {
    const line = lines[i].trim()

    if (line.startsWith('>')) {
      if (currentSeq.length > 0) {
        addSequence(currentSeq, currentSeqName, seqs, seqNames)
      }
      // eslint-disable-next-line unicorn/prefer-string-slice
      currentSeqName = line.substring(1, line.length)
      currentSeq = ''
    } else {
      // NOTE: Strip all characters except capital letters, asterisks, dots ans question marks
      const seq = line.toUpperCase().replace(/[^*.?A-Z]/g, '')
      currentSeq += seq
    }
  }

  if (currentSeq.length > 0) {
    addSequence(currentSeq, currentSeqName, seqs, seqNames)
  }

  return seqs
}

File: packages/algorithms/src/parseSequences.cpp
#include <fmt/format.h>

#include <boost/algorithm/string.hpp>
#include <map>
#include <regex>

namespace {
  using regex = std::regex;
  using std::regex_replace;
}// namespace

auto sanitizeLine(std::string line) {
  line = regex_replace(line, regex("\r\n"), "\n");
  line = regex_replace(line, regex("\r"), "\n");
  boost::trim(line);
  return line;
}

auto sanitizeSequence(std::string seq) {
  boost::to_upper(seq);
  // NOTE: Strip all characters except capital letters, asterisks, dots and question marks
  const auto re = regex("[^.?*A-Z]");
  seq = regex_replace(seq, re, "", std::regex_constants::match_any);
  return seq;
}

void addSequence(std::string currentSeqName, const std::string &currentSeq,
  std::map<std::string, std::string> &seqs,// NOLINT(google-runtime-references)
  std::map<std::string, int> &seqNames     // NOLINT(google-runtime-references)
) {
  if (currentSeqName.empty()) {
    currentSeqName = "Untitled";
  }

  auto it = seqNames.find(currentSeqName);
  if (it != seqNames.end()) {
    const auto nameCount = it->second;
    currentSeqName = fmt::format("{:s} ({:d})", currentSeqName, nameCount);
    it->second += 1;
  } else {
    seqNames.emplace(currentSeqName, 1);
  }

  const auto currentSeqSane = sanitizeSequence(currentSeq);
  seqs.emplace(currentSeqName, currentSeqSane);
}


std::map<std::string, std::string> parseSequences(std::istream &istream) {
  std::string currentSeqName;
  std::string currentSeq;
  std::map<std::string, std::string> seqs;
  std::map<std::string, int> seqNames;

  std::string line;
  while (std::getline(istream, line)) {
    line = sanitizeLine(line);

    if (boost::starts_with(line, ">")) {
      if (!currentSeq.empty()) {
        addSequence(currentSeqName, currentSeq, seqs, seqNames);
      }

      currentSeqName = line.substr(1, line.size());
      boost::trim(currentSeqName);
      currentSeq = "";
    } else {
      currentSeq += line;
    }
  }

  if (!currentSeq.empty()) {
    addSequence(currentSeqName, currentSeq, seqs, seqNames);
  }

  return seqs;
}
